scrooge-generator/src/main/scala/com/twitter/scrooge/backend/node/NodeGenerator.scala

```
@@ -0,0 +1,326 @@
+
+  private[this] object NodeGenerator {
+
+    object NodeKeywords {
+      private[this] val keywords = Set(
+        "break", "case", "catch", "class", "const", "continue", "debugger",
+        "default", "delete", "do", "else", "enum", "export", "extends", "false",
+        "finally", "for", "function", "if", "import", "in", "instanceof", "new",
+        "null", "return", "super", "switch", "this", "throw", "true", "try", "typeof",
+        "var", "void", "while", "with")
+
+      def contains(str: String): Boolean = keywords.contains(str.toLowerCase)
+    }
+
+  }
+
+  def quoteKeyword(str: String): String =
+    if (NodeKeywords.contains(str))
+      s"_$str"
+    else
+      str
+
+  def genType(t: FunctionType): CodeFragment = t match {
+    case bt: BaseType => genPrimitiveType(bt)
+    case Void => v("void")
+    case StructType(st, _) => v(s"${genID(st.sid.toTitleCase)}")
+    case EnumType(st, _) => v(s"${genID(st.sid.toTitleCase)}")
+    case MapType(k, vv, _) =>
+      v("Map<" + genType(k).toData + ", " + genType(vv).toData + ">")
+    case SetType(x, _) =>
+      v("Set<" + genType(x).toData + ">")
+    case ListType(x, _) =>
+      v("Array<" + genType(x).toData + ">")
+    case t: NamedType =>
+      val id = doc.qualifySimpleID(t.sid, namespaceLanguage, defaultNamespace)
+      v(genID(id.toTitleCase).toData)
+    case r: ReferenceType =>
+      throw new ScroogeInternalException("ReferenceType should not appear in backend")
+    case x =>
+      if (x.toString().equals("Void")) {
+        v("void")
+      } else {
+        v("any")
+      }
+  }
+
+
+  def genPrimitiveType(t: FunctionType): CodeFragment = t match {
+    case Void => v("void")
+    case TBool => v("boolean")
+    case TByte => v("byte")
+    case TDouble => v("number")
+    case TI16 => v("number")
+    case TI32 => v("number")
+    case TI64 => v("Int64")
+    case TString => v("string")
+    case TBinary => v("Buffer")
+    case _ => v("unknownPrimitive")
```

should this fail fast instead?